File: apps/shared/src/commonTest/kotlin/dev/supermux/ui/ActivityBodyTest.kt
package dev.supermux.ui

import dev.supermux.proto.ActivityToolBody
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertNotNull
import kotlin.test.assertNull
import kotlin.test.assertTrue

class ActivityBodyTest {
    @Test
    fun resolveBashParts_merges_start_and_result() {
        val parts = resolveBashParts(
            body = ActivityToolBody(kind = "bash", command = "ls -la"),
            resultBody = ActivityToolBody(kind = "bash", output = "a\nb", exitCode = 0),
            input = null,
            output = null,
            toolName = "Bash",
        )
        assertNotNull(parts)
        assertEquals("ls -la", parts.command)
        assertEquals("a\nb", parts.output)
        assertEquals(0, parts.exitCode)
    }

    @Test
    fun resolveEditParts_prefers_body_diff() {
        val parts = resolveEditParts(
            body = ActivityToolBody(kind = "edit", path = "a.ts", diff = "@@\n-a\n+b", oldText = "a", newText = "b"),
            input = null,
            toolName = "Edit",
        )
        assertNotNull(parts)
        assertEquals("a.ts", parts.path)
        assertTrue(parts.diff!!.contains("-a"))
    }

    @Test
    fun resolveEditParts_write_becomes_add_lines() {
        val parts = resolveEditParts(
            body = ActivityToolBody(kind = "write", path = "n.ts", content = "x"),
            input = null,
            toolName = "Write",
        )
        assertNotNull(parts)
        assertEquals("add", parts.mode)
        assertEquals("+x", parts.diff)
    }

    @Test
    fun resolveBashParts_null_for_non_bash() {
        assertNull(
            resolveBashParts(
                body = ActivityToolBody(kind = "generic", input = "x"),
                resultBody = null,
                input = "x",
                output = null,
                toolName = "Read",
            ),
        )
    }
}

File: src/web-app/src/lib/activity-body.test.ts
import { expect, test } from "bun:test"
import { resolveBashParts, resolveEditParts } from "./activity-body"

test("resolveBashParts merges start command + result output", () => {
  expect(resolveBashParts({
    body: { kind: "bash", command: "ls -la" },
    resultBody: { kind: "bash", output: "a\nb", exitCode: 0 },
    toolName: "Bash",
  })).toEqual({ command: "ls -la", output: "a\nb", exitCode: 0 })
})

test("resolveBashParts falls back to medium strings", () => {
  expect(resolveBashParts({
    input: "npm test",
    output: "ok",
    toolName: "Bash",
  })).toEqual({ command: "npm test", output: "ok", exitCode: undefined })
})

test("resolveEditParts prefers structured body", () => {
  const r = resolveEditParts({
    body: { kind: "edit", path: "a.ts", diff: "@@\n-a\n+b", oldText: "a", newText: "b" },
    toolName: "Edit",
  })
  expect(r).toMatchObject({ path: "a.ts", diff: "@@\n-a\n+b" })
})

test("resolveEditParts write body becomes add-style content", () => {
  const r = resolveEditParts({
    body: { kind: "write", path: "n.ts", content: "x" },
    toolName: "Write",
  })
  expect(r).toMatchObject({ path: "n.ts", mode: "add", content: "x" })
  expect(r!.diff).toContain("+x")
})
